Source/Session.swift

```diff
@@ -220,15 +220,20 @@ open class Session {
 
     // MARK: - DataRequest
 
+    public typealias RequestModifier = (inout URLRequest) throws -> Void
+
     struct RequestConvertible: URLRequestConvertible {
         let url: URLConvertible
         let method: HTTPMethod
         let parameters: Parameters?
         let encoding: ParameterEncoding
         let headers: HTTPHeaders?
+        let requestModifier: RequestModifier?
 
         func asURLRequest() throws -> URLRequest {
-            let request = try URLRequest(url: url, method: method, headers: headers)
+            var request = try URLRequest(url: url, method: method, headers: headers)
+            try requestModifier?(&request)
```

Comment:
This is exactly what I had envisioned @jshier. Looks solid so far with what you have. I don't really follow the reasoning of pointing out that you're not modifying the original `URLRequestConvertible` data. I don't really know why you would. Could you elaborate a bit more on what you're thinking there?

Reply:
I just mean that the modifiers only apply to the APIs which take the request components, not to the API which takes `URLRequestConvertible` values directly.

Reply:
Gotcha, that completely makes sense. If anyone is providing `URLRequestConvertible` values directly, they should create `requestModifier` types of their own. Exactly like I'm doing in my internal library.